ecs-agent/tcs/handler/handler.go

```diff
@@ -0,0 +1,258 @@
+	client.AddRequestHandler(ackPublishHealthMetricHandler(timer, session.heartbeatTimeout, session.heartbeatJitterMax))
+	client.AddRequestHandler(ackPublishInstanceStatusHandler(timer, session.heartbeatTimeout, session.heartbeatJitterMax))
+	client.SetAnyRequestHandler(anyMessageHandler(client, wsRWTimeout))
+	serveC := make(chan error, 1)
+	go func() {
+		serveC <- client.Serve(ctx)
+	}()
+	select {
+	case <-ctx.Done():
+		// outer context done, agent is exiting
+		client.Disconnect()
+	case <-timer.C:
+		seelog.Info("TCS Connection hasn't had any activity for too long; disconnecting")
+		client.Disconnect()
+	case err := <-serveC:
+		return err
+	}
+	return nil
+}
+
+// heartbeatHandler resets the heartbeat timer when HeartbeatMessage message is received from tcs.
+func heartbeatHandler(timer *time.Timer, heartbeatTimeout, heartbeatJitter time.Duration) func(*ecstcs.HeartbeatMessage) {
+	return func(*ecstcs.HeartbeatMessage) {
+		logger.Debug("Received HeartbeatMessage from tcs")
+		timer.Reset(retry.AddJitter(heartbeatTimeout, heartbeatJitter))
+	}
+}
+
+// ackPublishMetricHandler consumes the ack message from the backend. THe backend sends
+// the ack each time it processes a metric message.
+func ackPublishMetricHandler(timer *time.Timer, heartbeatTimeout, heartbeatJitter time.Duration) func(*ecstcs.AckPublishMetric) {
+	return func(*ecstcs.AckPublishMetric) {
+		logger.Debug("Received AckPublishMetric from tcs")
+		timer.Reset(retry.AddJitter(heartbeatTimeout, heartbeatJitter))
+	}
+}
+
+// ackPublishHealthMetricHandler consumes the ack message from backend. The backend sends
+// the ack each time it processes a health message
+func ackPublishHealthMetricHandler(timer *time.Timer, heartbeatTimeout, heartbeatJitter time.Duration) func(*ecstcs.AckPublishHealth) {
+	return func(*ecstcs.AckPublishHealth) {
+		logger.Debug("Received ACKPublishHealth from tcs")
+		timer.Reset(retry.AddJitter(heartbeatTimeout, heartbeatJitter))
+	}
+}
+
+// ackPublishInstanceStatusHandler consumes the ack message from backend. The backend sends
+// the ack each time it processes a health message
+func ackPublishInstanceStatusHandler(timer *time.Timer, heartbeatTimeout, heartbeatJitter time.Duration) func(*ecstcs.AckPublishInstanceStatus) {
+	return func(*ecstcs.AckPublishInstanceStatus) {
+		logger.Debug("Received AckPublishInstanceStatus from tcs")
+		timer.Reset(retry.AddJitter(heartbeatTimeout, heartbeatJitter))
+	}
+}
+
+// anyMessageHandler handles any server message. Any server message means the
+// connection is active
+func anyMessageHandler(client wsclient.ClientServer, wsRWTimeout time.Duration) func(interface{}) {
+	return func(interface{}) {
+		logger.Trace("TCS activity occurred")
+		// Reset read deadline as there's activity on the channel
+		if err := client.SetReadDeadline(time.Now().Add(wsRWTimeout)); err != nil {
+			logger.Warn("Unable to extend read deadline for TCS connection", logger.Fields{
+				field.Error: err,
+			})
+		}
```

Comment:
Just for my understanding, why do we reset the heartbeat timer in other handlers, but set the read deadline in this handler? Why do we not have a similar behavior?

Reply:
This is current implementation that I did not look deeply into. IIUC, heartbeat handlers and any message handlers serve different purpose (and seems to be double guarantee) - heartbeat handler is to reset the timer, which TCS client uses to actively DCed TCS client connection, while any message handler generally extend the allowed time we call read for the messages, and that is a required parameter from gorilla websocket connection. Specifically, `wsRWTimeout = 2*heartbeatTimeout + heartbeatJitterMax` and therefore should not supposed to be triggered but is necessary to be set so that the gorilla ws never rejects read before we actively do so. I think we can move this block to every request handler, but to reduce duplication and necessary error handling, the original implementation choose to implement it this way.

Reply:
Got it. Thank you for explaining.